src/ParseNode.js

```diff
@@ -315,11 +380,71 @@ export function assertNodeType<NODETYPE: NodeType>(
  * returns null.
  */
 export function checkNodeType<NODETYPE: NodeType>(
-    // The union allows either ParseNode<*> or the union of two specific nodes.
-    node: ?ParseNode<*> | ParseNode<*>,
+    node: ?AnyParseNode,
     type: NODETYPE,
 ): ?ParseNode<NODETYPE> {
-    return node && node.type === type ?
-        (node: ParseNode<NODETYPE>) :
-        null;
+    if (node && node.type === type) {
+        // $FlowFixMe: Inference not sophisticated enough to figure this out.
+        return node;
+    }
+    return null;
+}
+
+/**
+ * Returns the node more strictly typed iff it is of the given type. Otherwise,
+ * returns null.
+ */
+export function assertNodeTypes2<NODETYPE1: NodeType, NODETYPE2: NodeType>(
+    node: ?AnyParseNode,
+    type1: NODETYPE1,
+    type2: NODETYPE2,
+): ParseNode<NODETYPE1> | ParseNode<NODETYPE2> {
+    const typedNode = checkNodeTypes2(node, type1, type2);
+    if (!typedNode) {
+        throw new Error(
+            `Expected node of type ${type1} or ${type2}, but got ` +
+            (node ? `node of type ${node.type}` : String(node)));
+    }
+    // $FlowFixMe: Might be related to https://github.com/facebook/flow/issues/6379
+    return typedNode;
+}
+
+/**
+ * Returns the node more strictly typed iff it is of the given type. Otherwise,
+ * returns null.
+ */
+export function checkNodeTypes2<NODETYPE1: NodeType, NODETYPE2: NodeType>(
+    node: ?AnyParseNode,
+    type1: NODETYPE1,
+    type2: NODETYPE2,
+): ?ParseNode<NODETYPE1> | ParseNode<NODETYPE2> {
+    // Explicit type needed due to https://github.com/facebook/flow/issues/6379
+    return checkNodeType(node, type1) ||
+        (checkNodeType(node, type2): ?ParseNode<NODETYPE2>);
+}
+
+/**
+ * Returns the node more strictly typed iff it is of the given type. Otherwise,
+ * returns null.
+ */
+export function assertSymbolNodeType(node: ?AnyParseNode): SymbolParseNode {
+    const typedNode = checkSymbolNodeType(node);
+    if (!typedNode) {
+        throw new Error(
+            `Expected node of symbol group type, but got ` +
+            (node ? `node of type ${node.type}` : String(node)));
+    }
+    return typedNode;
+}
+
+/**
+ * Returns the node more strictly typed iff it is of the given type. Otherwise,
+ * returns null.
+ */
+export function checkSymbolNodeType(node: ?AnyParseNode): ?SymbolParseNode {
+    if (node && GROUPS.hasOwnProperty(node.type)) {
```

Comment:
We'll be able to get rid of this weirdness once we have a separate `atom` parse node type.

Reply:
Yes! This is exactly what I was referring to when you originally suggested the `atom` idea.